Move services to subpath for development setup

BacLuc opened the pull request:
Like this its possible to access the local instance from the mobile phone if you are in the same network, or to use
github codespaces to run ecamp.

Needs https://github.com/ecamp/hal-json-vuex/pull/280

- [x] test with github codespace and mobile
- [x] access local instance from mobile phone in the same network
- [x] Fix react print, some uri's still have the baseUrl and are "not preloaded" (print react worked fine without errors, don't know what happened here. the error did not reappier)
- [x] Add new callback urls
  - [x] add http://localhost:3000/api/auth/pbsmidata/callback to pbs.puzzle.ch
  - [x] add http://localhost:3000/api/auth/cevidb/callback to cevi.puzzle.ch (login is not really public, so I just changed it without adding to keepass)
  - [x] add http://localhost:3000/api/auth/jubladb/callback to jubla.puzzle.ch (login is not really public, so I just changed it without adding to keepass)
  - [x] to google oauth (i did not find the oauth2 config in the shared google account)

BacLuc committed:
cosmetics api/config/packages/framework.yaml: convert array syntax

That it can easily be seen what will be added.
And that the line does not get too long.

BacLuc committed:
dev setup: put services behind reverseproxy

That they are all available on the same port.
This makes it possible to debug the application locally,
on a github codespace or with ngrok.

Issue: #3214

BacLuc commented on api/docker/caddy/Caddyfile:
```diff
@@ -2,49 +2,71 @@
 
-log
+	handle_path /api* {
+	    header X-Forwarded-Prefix /api
```
This is the magic, with this header the api can handle it perfectly that it's behind a subpath of a reverse proxy.

manuelmeister replied:
Leave comment, this could be a small performance drop

BacLuc replied:
done

BacLuc commented on api/docker/caddy/Caddyfile:
```diff
@@ -2,49 +2,71 @@
     # Debug
```
TODO: format Caddyfile

BacLuc replied:
done

BacLuc commented on api/docker/caddy/Caddyfile:
```diff
@@ -2,49 +2,71 @@
-    encode zstd gzip
-    file_server
+        php_fastcgi unix//var/run/php/php-fpm.sock {
+            env HTTP_X_FORWARDED_PREFIX /api
```
oh oh, i hope i can remove this

BacLuc replied:
no, id did not work with the header directive

BacLuc commented on docker-compose.yml:
```diff
@@ -57,6 +56,9 @@ services:
       # This should correspond to the server declared in PHPStorm `Preferences | Languages & Frameworks | PHP | Servers`
       # Then PHPStorm will use the corresponding path mappings
       PHP_IDE_CONFIG: serverName=localhost
+      CORS_ALLOW_ORIGIN: "*"
+      API_DOMAIN: ".*"
+      SHARED_COOKIE_DOMAIN: ""
```
we have to discuss how we want to handle the security for the development environment

manuelmeister commented:
How does this work?

manuelmeister commented:
## Core Meeting Decision

We want this. Maybe for prod as well. @usu said he read somewhere that this has some performance gains.
Edit callback urls of oauth

usu commented:
> ## Core Meeting Decision
> We want this. Maybe for prod as well. @usu said he read somewhere that this has some performance gains.

Found the article again from Kevin Dunglas on his personal blog: https://dunglas.dev/2022/01/preventing-cors-preflight-requests-using-content-negotiation/

pmattmann commented:
OAuth is registered in ecamp3development@gmail.com
![image](https://user-images.githubusercontent.com/470237/216824570-d9000f97-5b5f-4fca-a771-0e282a666e6a.png)

I added password to the tresor.
I added callback to OAuth-config; OAuth with google now working.

usu commented:
> OAuth is registered in [ecamp3development@gmail.com](mailto:ecamp3development@gmail.com) ![image](https://user-images.githubusercontent.com/470237/216824570-d9000f97-5b5f-4fca-a771-0e282a666e6a.png)
> 
> I added password to the tresor. I added callback to OAuth-config; OAuth with google now working.

And we use ecamp3core for the production OAuth? Or what was the purpose to use ecamp3development?

BacLuc commented:
I would propose that we merge this PR, before the callback uris are added.

usu commented on frontend/public/environment.dist:
```diff
@@ -1,7 +1,7 @@
 window.environment = {
-  API_ROOT_URL: 'http://localhost:3001',
+  API_ROOT_URL: '/api',
```
Wouldn't `API_ROOT_URL: 'http://localhost:3000/api'` be a bit more explicit?

BacLuc replied:
Then you cannot access your local instance from your phone,
or the instance running on github codespaces.

usu replied:
Ah, makes sense. I didn't test from mobile yet, still need to figure out how to bypass firewall and stuff

usu commented on frontend/public/environment.dist:
```diff
@@ -1,7 +1,7 @@
+  API_ROOT_URL: '/api',
   COOKIE_PREFIX: 'localhost_',
-  PRINT_URL: 'http://localhost:3003',
+  PRINT_URL: '../print',
```
This doesn't work. Should be `'/print'` or `'http://localhost:3000/print'`

BacLuc replied:
Then you cannot use print your local instance from your phone,
or the instance running on github codespaces.
Sadly we only have one axios instance which we use for hal-json-vuex.

And we currently use hal-json-vuex to fetch the pdf, even though we don't want to store the pdf in the vuex store.
Will check if the print works

BacLuc replied:
I tested it again on gitpod.
Print uses the internal hostname of the api (inside the docker-compose network), this works.
I made a snapshot of the gitpod, you can try it yourself: https://gitpod.io#snapshot/d6ecd9af-2148-44ac-b593-89dcbcb03f70
But it takes a while, the pod this is running in does not have a lot of resources.

usu replied:
With `'../print'` this is my print preview (basically an infinite loop of iframes):

<img width="637" alt="image" src="https://user-images.githubusercontent.com/449555/218850059-67ba2a80-3823-4c14-8231-86fe56455979.png">

usu replied:
> And we currently use hal-json-vuex to fetch the pdf, even though we don't want to store the pdf in the vuex store.

Hmm, I don't think so. We fetch the pdf with the normal axios instance, or did I misunderstood something?
https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/print/print-nuxt/generatePdfMixin.js#L34

If we would fetch the pdf with hal-json-vuex, that would probably not go well. I've stumbled on this at a different place already: https://github.com/ecamp/hal-json-vuex/issues/278

BacLuc replied:
Hmm, I don't think so. We fetch the pdf with the normal axios instance, or did I misunderstood something?
https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/print/print-nuxt/generatePdfMixin.js#L34

Yes, you are right.

I fixed the print preview (and the nuxt print) in https://github.com/ecamp/ecamp3/pull/3222/commits/4f65c7d64396d74542839d919425be28ef16f578

usu reviewed (commented):
Printing with nuxt doesn't work for me. I guess `print/print.env` needs some adjustments as well.

carlobeltrame commented:
@BacLuc I added the OAuth callback URLs to the cevi and jubla integration servers

BacLuc committed:
print: use same path for the api as the frontend

That when we put the fastcgi directive right where we handle the /api subpath,
it will work too.

usu reviewed (approved):
Nice, printing works now for me.

I think one of the last commits broke the OpenAPI docs. When hitting http://localhost:3000/api, all the static assets are not loaded. I thought this worked before (but maybe I'm mistaken and it was only working on http://localhost:3001).

One thing we might want to consider: After merging this PR, all developers need to delete the environment.js (or update manually). This doesn't scale well, once more people are working on this project. So we might want to find a smarter way of not creating a hardcopy of the environment.js during startup. Will open a separate ticket for this. (https://github.com/ecamp/ecamp3/issues/3267)

carlobeltrame commented on api/docker/caddy/Caddyfile:
```diff
@@ -1,50 +1,76 @@
-    file_server
+:3000 {
+	log {
+		level DEBUG
```
Is it a good idea to set DEBUG log level directly in the Caddyfile? This will be used also in production, won't it? Did you intend to remove this before merging?

BacLuc replied:
No, for that we have caddyfile.prod

carlobeltrame commented on api/docker/caddy/Caddyfile:
```diff
@@ -1,50 +1,76 @@
+		level DEBUG
+	}
+
+	handle_path /api* {
```
Won't this also attempt to handle requests to `/apitest/1234`? And same for `/printxyz` and `/mailer/foo`?

BacLuc replied:
Yes

carlobeltrame commented on api/docker/caddy/Caddyfile:
```diff
@@ -1,50 +1,76 @@
+	}
+
+	handle_path /print* {
+		reverse_proxy print:3003
```
How about setting the X-Forwarded-Prefix for these other services as well? I could easily imagine the print or mail (or even frontend) service outputting an absolute link to another print or mail or frontend page (e.g. when serving an error response, the print service might link to some page with further info / logs, or the mail service might use links for navigation). For generating the correct hrefs, these services need to know the base URL under which they appear in the public.
I see that you tell the mail service about this via the MH_UI_WEB_PATH env variable. But it still wouldn't hurt to set the header here, in case we ever switch to a different dummy mail server.

BacLuc replied:
We could do it with print, but the print service currently does not use the header as far as i know.

The mail service gets the requests with /mail/.. (Thats the difference between handle_path and handle, handle does not rewrite the request).
mailhog does not use urls relative to the request url for its assets, that's why you have to tell mailhog under which path its reachable. If the request is not rewritten, the X-Forwared-Prefix Header does not make sense as far as i understand the header.

carlobeltrame commented on docker-compose.yml:
```diff
@@ -57,6 +56,9 @@ services:
       # This should correspond to the server declared in PHPStorm `Preferences | Languages & Frameworks | PHP | Servers`
       # Then PHPStorm will use the corresponding path mappings
```
Are the path mappings still correct?

carlobeltrame commented on docker-compose.yml:
```diff
@@ -80,18 +82,14 @@ services:
+      - target: 3000
+        published: 3000
+        protocol: tcp
+      # legacy api until oauth callback url are changed
       - target: 3001
         published: 3001
         protocol: tcp
```
Remove, now that the callback URLs are changed?

carlobeltrame commented on docker-compose.yml:
```diff
@@ -80,18 +82,14 @@ services:
       MERCURE_EXTRA_DIRECTIVES: demo
     ports:
-      # HTTP
+      # frontend
```
Isn't this comment wrong? All services are served via port 3000 now, not just the frontend.

carlobeltrame commented on docker-compose.yml:
```diff
@@ -57,6 +56,9 @@ services:
       # This should correspond to the server declared in PHPStorm `Preferences | Languages & Frameworks | PHP | Servers`
       # Then PHPStorm will use the corresponding path mappings
       PHP_IDE_CONFIG: serverName=localhost
+      CORS_ALLOW_ORIGIN: "*"
+      API_DOMAIN: ".*"
+      SHARED_COOKIE_DOMAIN: ""
```
I don't understand why you chose these values.
- CORS_ALLOW_ORIGIN could be empty now, couldn't it? There are no more cross-origin requests after this PR.
- API_DOMAIN is a regex now, which makes no sense. If this is really needed (I assume you did this for accessing the API from a mobile phone), could you use another environment variable such as `ADDITIONAL_TRUSTED_HOSTS` or `TRUSTED_API_HOSTS` or something similar? That way it would become clearer that the API is not running on a regex domain.
- If we follow this PR up with #3262, then SHARED_COOKIE_DOMAIN can be removed

BacLuc replied:
> I don't understand why you chose these values.
> 
>     * CORS_ALLOW_ORIGIN could be empty now, couldn't it? There are no more cross-origin requests after this PR.
> 
Maybe, i did not test it. But framework.yml adds localhost and and caddy as CORS Hosts, maybe it does not work if you don't include the same host explicitly.
The goal of this restructuring is that you can access the instance on your local machine from your phone in the same network. (And but with lower priority an instance running on github codespaces or gitpod)
>     * API_DOMAIN is a regex now, which makes no sense. If this is really needed (I assume you did this for accessing the API from a mobile phone), could you use another environment variable such as `ADDITIONAL_TRUSTED_HOSTS` or `TRUSTED_API_HOSTS` or something similar? That way it would become clearer that the API is not running on a regex domain.
I can do that
> 
>     * If we follow this PR up with [.helm: move services to subpaths #3262](https://github.com/ecamp/ecamp3/pull/3262), then SHARED_COOKIE_DOMAIN can be removed

SHARED_COOKIE_DOMAIN is removed in https://github.com/ecamp/ecamp3/pull/3262/files#diff-09c8c391e3a627b701028c8c82756440688ff60ec3fe6664b93f4a92e68e5df0

BacLuc replied:
Tested empty CORS_ALLOW_ORIGIN and removed it in https://github.com/ecamp/ecamp3/pull/3222/commits/cf9d20fd1805a3fdb745f50b635b72f2570c8529
Renamed API_DOMAIN in https://github.com/ecamp/ecamp3/pull/3222/commits/dc65e65dcc5095a84b08c5570c76af5178ec845e

carlobeltrame commented on frontend/src/plugins/__tests__/auth.spec.js:
```diff
@@ -31,6 +31,8 @@ const validJWTPayload =
 //   "user": "/users/1a2b3c4d"
 // }
 
+const envBackup = JSON.parse(JSON.stringify(window.environment))
```
Consider creating a function `deepClone` or similar which explains why you are using `JSON.parse(JSON.stringify())` in this file, or use the lodash `clone` function instead. I was confused when first reading this.

BacLuc replied:
fixed in https://github.com/ecamp/ecamp3/pull/3222/commits/775f63269c2a2670567c089b5976f2b89f301310

carlobeltrame commented on frontend/src/plugins/auth.js:
```diff
@@ -113,10 +113,20 @@ async function redirectToOAuthLogin(provider) {
     .href(apiStore.get(), provider, { callback: encodeURI(returnUrl) })
     .then((url) => {
-      window.location.href = window.environment.API_ROOT_URL + url
+      const apiRootUrl = window.environment.API_ROOT_URL
+      window.location.href = (hasHost(apiRootUrl) ? apiRootUrl : '') + url
```
This seems very much specific to our current setup of "either all services live on a subdomain with no path prefix, or they all live on the same domain with different path prefixes". In the rest of this PR, I was under the impression that everything is generalized so that services may live on subdomains or have path prefixes, or both.
Isn't this more or less a bug in hal-json-vuex? I feel like hal-json-vuex should either return absolute URIs, or cut off the path prefix from all URIs. What is your opinion on this?

BacLuc replied:
hal-json-vuex does everything right i think.
We have a mixture between using hal-json-vuex and the browser.
Haljsonvuex uses relative paths from apiRootUrl, and adds the prefix before the requests, but window.location.href does not know about this.
Thus the logic to put together urls is duplicated.

Before #3262 , we still use an api url with a different host, so we have to include the host when putting together the url.
After (or in) #3262, we can remove this and use relative url's everywhere.

carlobeltrame commented on print/print.env:
```diff
@@ -1,7 +1,7 @@
 PRINT_URL=http://print:3003
-COOKIE_PREFIX=localhost_
\ No newline at end of file
+COOKIE_PREFIX=localhost_
```
I think in #3262 you can also remove this variable. Or am I overlooking something?

BacLuc committed:
refactor/api: rename env variable API_DOMAIN to ADDITIONAL_TRUSTED_HOSTS

That it's clear for what the variable is used and that it can contain a regex
in the development setup.

Issue: #3214

usu commented:
The end2ned tests currently fail (=timeout) because `wait-for-container-startup.sh` still probes for localhost:3001 which doesn't respond anymore. Or is port 3001 still expected to work? I see there are still references to 3001 in the Caddyfile and being a bit confused of whether this is as intended or code that still needs to be converted.

carlobeltrame commented:
It needs to be changed. 3001 is not exposed outside of the caddy container anymore, it's just used internally for redirecting traffic to the correct service. Browser -> caddy:3000/api -> caddy:3000 -> php

BacLuc commented:
> The end2ned tests currently fail (=timeout) because `wait-for-container-startup.sh` still probes for localhost:3001 which doesn't respond anymore. Or is port 3001 still expected to work? I see there are still references to 3001 in the Caddyfile and being a bit confused of whether this is as intended or code that still needs to be converted.

Yes, i had to fix wait-for-container-startup.sh
